crates/mun_memory/src/mapping.rs

```diff
@@ -1,35 +1,71 @@
+use abi::Guid;
+
 use crate::{
-    diff::{diff, Diff, FieldDiff, FieldEditKind},
+    diff::{diff, Diff, FieldDiff},
     gc::GcPtr,
     r#type::Type,
+    ArrayType, TypeKind,
 };
 use std::collections::{HashMap, HashSet};
 
 pub struct Mapping {
     pub deletions: HashSet<Type>,
-    pub conversions: HashMap<Type, Conversion>,
+    pub struct_mappings: HashMap<Type, StructMapping>,
     pub identical: Vec<(Type, Type)>,
 }
 
-pub struct Conversion {
+pub struct StructMapping {
     pub field_mapping: Vec<FieldMapping>,
     pub new_ty: Type,
 }
 
 /// Description of the mapping of a single field. When stored together with the new index, this
 /// provides all information necessary for a mapping function.
+#[derive(Debug)]
 pub struct FieldMapping {
     pub new_ty: Type,
     pub new_offset: usize,
     pub action: Action,
 }
 
 /// The `Action` to take when mapping memory from A to B.
-#[derive(Eq, PartialEq)]
+#[derive(Debug, Eq, PartialEq)]
 pub enum Action {
-    Cast { old_offset: usize, old_ty: Type },
-    Copy { old_offset: usize },
-    Insert,
+    /// Allocate a new array.
+    ArrayAlloc,
+    /// Allocate a new array and initialize it with a single value.
+    ArrayFromValue {
+        element_action: Box<Action>,
+        old_offset: usize,
+    },
+    /// Allocate a new array and map values from an old array.
+    ArrayMap {
+        element_action: Box<Action>,
+        old_offset: usize,
+    },
+    /// Cast a primitive type.
+    Cast { old_ty: Type, old_offset: usize },
+    /// Copy bytes.
+    Copy {
+        old_offset: usize,
+        /// Size in bytes
+        size: usize,
+    },
+    /// Replace an array with its element type, copying its first element - if any.
+    ElementFromArray {
+        element_action: Box<Action>,
+        old_offset: usize,
+    },
+    /// Allocate a new struct and ensure zero-initalization.
+    StructAlloc,
+    /// Allocate a new struct and map from a heap-allocated struct.
+    StructMapFromGc { old_ty: Type, old_offset: usize },
+    /// Allocate a new struct and map from a value struct.
+    StructMapFromValue { old_ty: Type, old_offset: usize },
+    /// Map a value struct in-place.
+    StructMapInPlace { old_ty: Type, old_offset: usize },
+    /// Ensure the memory is zero-initialized.
+    ZeroInitialize,
 }
 
 impl Mapping {
```

Comment:
Are there any tests that verify the correctness of the Mapping? Besides the `diff` tests?

Reply:
Only the integration tests in the `mun_runtime`